Add a CLAUDE.md file for ofrak. (#684)
redballoonsecurity/ofrak: CLAUDE.md

# CLAUDE.md

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## Project Overview

OFRAK (Open Firmware Reverse Analysis Konsole) is a binary analysis and modification platform for reverse engineering firmware, executables, and file formats. See `README.md` for full details.

## Quick Reference Commands

```bash
# Development setup (core packages only - faster)
make develop-core

# Full setup including optional disassemblers
make develop

# Run tests for a package
make -C ofrak_core test

# Run single test
python3 -m pytest tests/path/to/test.py::TestClass::test_method -v

# Linting (from repo root)
make inspect                    # autoflake + black check
pre-commit run --all-files      # all hooks

# Type checking (from package directory)

> mypy

Reviewer: I would recommend `python3 -m mypy`. Just had Claude completely confused because of a venv created with `--system-site-packages`. `mypy` would result in use of `mypy` without venv, but `python3 -m mypy` would use venv

  - Reference asset files by path (e.g., `tests/components/assets/sample.bin`)
  - Instruct user to place real test files: "Place test file at `tests/components/assets/sample.bin`"
  - **Remind user**: Test data must be suitable for public distribution (self-created, public domain, or permissively licensed)
  - Don't create synthetic data in test code
  - Don't generate test files programmatically

When creating tests, instruct the user:

```
Please place a real MyFormat test file at:
  tests/components/assets/sample.myformat

IMPORTANT: Test data must be suitable for public distribution, as the OFRAK
repository is open source and publicly accessible. Use test data that is:
- Created by you
- Public domain
- Permissively licensed
- Otherwise freely redistributable

You can obtain appropriate test data by:
- Creating your own test file with [tool]
- Using public domain samples from [source]
- Generating test cases yourself
- Extracting from openly licensed firmware/archives
```

See `docs/contributor-guide/getting-started.md` for full coding standards and docstring format.

## Contributing

- Update `CHANGELOG.md` and bump the **rc version** in `setup.py` for any package changes

> - Changelogs follow Keep a Changelog format with PR links

Reviewer 2: Link to "Keep a Changelog" here? It's written as a proper noun, but deserves a reference, I think.

- Pre-commit hooks must pass (install with `pre-commit install`)

> See `CONTRIBUTING.md` for PR guidelines and maintainer info.

Reviewer 2: This should also link to the contributor guidelines in the docs folder. Even though the `CONTRIBUTING.md` has some important info, it is more relevant for real humans on GitHub than it is for agents using this file.

> See [CONTRIBUTING.md](./CONTRIBUTING.md) for PR guidelines and maintainer info.

Author: ```suggestion
See [CONTRIBUTING.md](./CONTRIBUTING.md) for PR guidelines and maintainer info.

See [Getting Started](docs/contributor-guide/getting-started.md) for full coding standards and docstring format.
```

# CLAUDE.md

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## Project Overview

OFRAK (Open Firmware Reverse Analysis Konsole) is a binary analysis and modification platform for reverse engineering firmware, executables, and file formats. See `README.md` for full details.

## Quick Reference Commands

```bash
# Development setup (core packages only - faster)
make develop-core

# Full setup including optional disassemblers
make develop

# Run tests for a package
make -C ofrak_core test

# Run single test
python3 -m pytest tests/path/to/test.py::TestClass::test_method -v

# Linting (from repo root)
make inspect                    # autoflake + black check
pre-commit run --all-files      # all hooks

# Type checking (from package directory)
mypy

# Build frontend
cd frontend && npm install && npm run build

# Procure an OFRAK license, if one does not exist already
# **CRITICAL**: you are **not** authorized to accept OFRAK license on user's behalf
# **MUST** tell the user to run the command themselves and to review OFRAK license
ofrak license

> ```

Author: ```suggestion
```

## Architecture

- **Entry point**: `OFRAK` class → `discover()` backends → `run()` or `create_ofrak_context()`
- **Example scripts**: See `examples/` directory and [tutorial documentation](https://ofrak.com/docs/getting-started/basic-usage.html) for basic OFRAK usage patterns
- **Core abstraction**: `Resource` - represents binary data with tags, supports `unpack()`, `run()`, `get_data()`, `flush_data_to_disk()`
- **Component types**: `Analyzer`, `Modifier`, `Unpacker`, `Packer`, `Identifier` (in `ofrak/component/`)
- **Resource lifecycle**: create → identify → unpack → analyze/modify → pack → flush
- **Dependency injection framework**: `synthol` - components discovered via `ofrak.packages` entry points
- **Async**: All I/O is async; tests use `asyncio_mode="auto"`

Key directories in `ofrak_core/src/ofrak/`:
- `component/` - Base component classes
- `core/` - Format implementations (elf/, pe/, compression, filesystems, firmware)
- `model/` - Data models (Resource, ResourceTag, views)
- `service/` - Core services (data, resource, job, component locator)

Frontend:

> - `frontend/` - Web GUI built with React/TypeScript. Build with `npm install && npm run build`

Author: ```suggestion
- `frontend/` - Web GUI built with Svelte 4 and Vite. Build with `npm install && npm run build`
```

# CLAUDE.md

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## Project Overview

OFRAK (Open Firmware Reverse Analysis Konsole) is a binary analysis and modification platform for reverse engineering firmware, executables, and file formats. See `README.md` for full details.

## Quick Reference Commands

```bash
# Development setup (core packages only - faster)
make develop-core

# Full setup including optional disassemblers
make develop

# Run tests for a package (from package directory)
cd ofrak_core && make test

# Run single test
python3 -m pytest tests/path/to/test.py::TestClass::test_method -v

# Linting (from repo root)
make inspect                    # autoflake + black check
pre-commit run --all-files      # all hooks

# Type checking (from package directory)
mypy

# Build frontend
cd frontend && npm install && npm run build

> ```

Author: At this stage we don't want to slow down adoption. Adoption != compliance.

## Package Structure

Packages must be installed in dependency order:
1. `ofrak_type` - Primitive types (enums, Range)
2. `ofrak_io` - I/O utilities

> 3. `ofrak_patch_maker` - Binary patching toolchains

Author: @claude can you address this
